File: bindings/csharp/SatoxSDK.cs
/**
 * @file SatoxSDK.cs
 * @brief Main C# binding for the Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

using System;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace Satox
{
    /// <summary>
    /// Main SDK class for Satox blockchain operations
    /// </summary>
    public class SatoxSDK : IDisposable
    {
        private CoreManager coreManager;
        private WalletManager walletManager;
        private SecurityManager securityManager;
        private AssetManager assetManager;
        private NFTManager nftManager;
        private BlockchainManager blockchainManager;
        private IPFSManager ipfsManager;
        private NetworkManager networkManager;
        private DatabaseManager databaseManager;
        
        private SDKConfig config;
        private bool isRunning;
        private readonly object lockObject = new object();
        
        /// <summary>
        /// Default constructor with default configuration
        /// </summary>
        public SatoxSDK() : this(new SDKConfig())
        {
        }
        
        /// <summary>
        /// Constructor with custom configuration
        /// </summary>
        /// <param name="config">The SDK configuration</param>
        public SatoxSDK(SDKConfig config)
        {
            this.config = config ?? new SDKConfig();
            this.isRunning = false;
        }
        
        /// <summary>
        /// Initialize the SDK with the current configuration
        /// </summary>
        /// <returns>Task that completes when initialization is done</returns>
        public async Task InitializeAsync()
        {
            lock (lockObject)
            {
                if (isRunning)
                {
                    throw new InvalidOperationException("SDK is already running");
                }
            }
            
            // Initialize core manager
            coreManager = new CoreManager();
            await coreManager.InitializeAsync(config);
            
            // Initialize wallet manager
            walletManager = new WalletManager();
            await walletManager.InitializeAsync(config);
            
            // Initialize security manager
            securityManager = new SecurityManager();
            await securityManager.InitializeAsync(config);
            
            // Initialize asset manager
            assetManager = new AssetManager();
            await assetManager.InitializeAsync(config);
            
            // Initialize NFT manager
            nftManager = new NFTManager();
            await nftManager.InitializeAsync(config);
            
            // Initialize blockchain manager
            blockchainManager = new BlockchainManager();
            await blockchainManager.InitializeAsync(config);
            
            // Initialize IPFS manager
            ipfsManager = new IPFSManager();
            await ipfsManager.InitializeAsync(config);
            
            // Initialize network manager
            networkManager = new NetworkManager();
            await networkManager.InitializeAsync(config);
            
            // Initialize database manager
            databaseManager = new DatabaseManager();
            await databaseManager.InitializeAsync(config);
            
            lock (lockObject)
            {
                isRunning = true;
            }
        }
        
        /// <summary>
        /// Start all SDK services
        /// </summary>
        /// <returns>Task that completes when all services are started</returns>
        public async Task StartAsync()
        {
            lock (lockObject)
            {
                if (!isRunning)
                {
                    throw new InvalidOperationException("SDK is not initialized");
                }
            }
            
            // Start all managers
            await coreManager.StartAsync();
            await walletManager.StartAsync();
            await securityManager.StartAsync();
            await assetManager.StartAsync();
            await nftManager.StartAsync();
            await blockchainManager.StartAsync();
            await ipfsManager.StartAsync();
            await networkManager.StartAsync();
            await databaseManager.StartAsync();
        }
        
        /// <summary>
        /// Stop all SDK services
        /// </summary>
        /// <returns>Task that completes when all services are stopped</returns>
        public async Task StopAsync()
        {
            lock (lockObject)
            {
                if (!isRunning)
                {
                    return;
                }
            }
            
            // Stop all managers
            if (coreManager != null) await coreManager.StopAsync();
            if (walletManager != null) await walletManager.StopAsync();
            if (securityManager != null) await securityManager.StopAsync();
            if (assetManager != null) await assetManager.StopAsync();
            if (nftManager != null) await nftManager.StopAsync();
            if (blockchainManager != null) await blockchainManager.StopAsync();
            if (ipfsManager != null) await ipfsManager.StopAsync();
            if (networkManager != null) await networkManager.StopAsync();
            if (databaseManager != null) await databaseManager.StopAsync();
            
            lock (lockObject)
            {
                isRunning = false;
            }
        }
        
        /// <summary>
        /// Connect to the blockchain network
        /// </summary>
        /// <returns>Task that completes when connected</returns>
        public async Task ConnectAsync()
        {
            lock (lockObject)
            {
                if (!isRunning)
                {
                    throw new InvalidOperationException("SDK is not running");
                }
            }
            await networkManager.ConnectAsync();
        }
        
        /// <summary>
        /// Disconnect from the blockchain network
        /// </summary>
        /// <returns>Task that completes when disconnected</returns>
        public async Task DisconnectAsync()
        {
            lock (lockObject)
            {
                if (!isRunning)
                {
                    return;
                }
            }
            await networkManager.DisconnectAsync();
        }
        
        /// <summary>
        /// Check if the SDK is running
        /// </summary>
        /// <returns>true if running, false otherwise</returns>
        public bool IsRunning
        {
            get
            {
                lock (lockObject)
                {
                    return isRunning;
                }
            }
        }
        
        /// <summary>
        /// Get the core manager
        /// </summary>
        public CoreManager CoreManager => coreManager;
        
        /// <summary>
        /// Get the wallet manager
        /// </summary>
        public WalletManager WalletManager => walletManager;
        
        /// <summary>
        /// Get the security manager
        /// </summary>
        public SecurityManager SecurityManager => securityManager;
        
        /// <summary>
        /// Get the asset manager
        /// </summary>
        public AssetManager AssetManager => assetManager;
        
        /// <summary>
        /// Get the NFT manager
        /// </summary>
        public NFTManager NFTManager => nftManager;
        
        /// <summary>
        /// Get the blockchain manager
        /// </summary>
        public BlockchainManager BlockchainManager => blockchainManager;
        
        /// <summary>
        /// Get the IPFS manager
        /// </summary>
        public IPFSManager IPFSManager => ipfsManager;
        
        /// <summary>
        /// Get the network manager
        /// </summary>
        public NetworkManager NetworkManager => networkManager;
        
        /// <summary>
        /// Get the database manager
        /// </summary>
        public DatabaseManager DatabaseManager => databaseManager;
        
        /// <summary>
        /// Get the current configuration
        /// </summary>
        public SDKConfig Config => config;
        
        /// <summary>
        /// Get the SDK version
        /// </summary>
        public static string Version => "1.0.0";
        
        /// <summary>
        /// Get the SDK build date
        /// </summary>
        public static string BuildDate => "2025-06-30";
        
        /// <summary>
        /// Dispose the SDK
        /// </summary>
        public void Dispose()
        {
            StopAsync().Wait();
        }
        
        /// <summary>
        /// Shutdown the SDK gracefully
        /// </summary>
        /// <returns>Task that completes when shutdown is done</returns>
        public async Task ShutdownAsync()
        {
            await StopAsync();
        }
    }
    
    /// <summary>
    /// SDK configuration class
    /// </summary>
    public class SDKConfig
    {
        public string DataDir { get; set; } = "./data";
        public string Network { get; set; } = "mainnet";
        public bool EnableMining { get; set; } = false;
        public bool EnableSync { get; set; } = true;
        public uint SyncIntervalMs { get; set; } = 1000;
        public uint MiningThreads { get; set; } = 1;
        public uint MaxConnections { get; set; } = 10;
        public ushort RPCPort { get; set; } = 7777;
        public ushort P2PPort { get; set; } = 67777;
        public bool EnableRPC { get; set; } = true;
        public bool EnableP2P { get; set; } = true;
        public string LogLevel { get; set; } = "info";
        public string LogFile { get; set; } = "./logs/satox-sdk.log";
    }
    
    /// <summary>
    /// Core manager class
    /// </summary>
    public class CoreManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Wallet manager class
    /// </summary>
    public class WalletManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Security manager class
    /// </summary>
    public class SecurityManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Asset manager class
    /// </summary>
    public class AssetManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// NFT manager class
    /// </summary>
    public class NFTManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Blockchain manager class
    /// </summary>
    public class BlockchainManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// IPFS manager class
    /// </summary>
    public class IPFSManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Network manager class
    /// </summary>
    public class NetworkManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
        public Task ConnectAsync() => Task.CompletedTask;
        public Task DisconnectAsync() => Task.CompletedTask;
    }
    
    /// <summary>
    /// Database manager class
    /// </summary>
    public class DatabaseManager
    {
        public Task InitializeAsync(SDKConfig config) => Task.CompletedTask;
        public Task StartAsync() => Task.CompletedTask;
        public Task StopAsync() => Task.CompletedTask;
    }
} 

File: bindings/go/satox_sdk.go
/**
 * @file satox_sdk.go
 * @brief Main Go binding for the Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

package satox

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// SatoxSDK represents the main SDK instance
type SatoxSDK struct {
	coreManager     *CoreManager
	walletManager   *WalletManager
	securityManager *SecurityManager
	assetManager    *AssetManager
	nftManager      *NFTManager
	blockchainManager *BlockchainManager
	ipfsManager     *IPFSManager
	networkManager  *NetworkManager
	databaseManager *DatabaseManager
	
	config     *Config
	isRunning  bool
	mutex      sync.RWMutex
	ctx        context.Context
	cancel     context.CancelFunc
}

// Config represents the SDK configuration
type Config struct {
	DataDir         string `json:"data_dir"`
	Network         string `json:"network"`
	EnableMining    bool   `json:"enable_mining"`
	EnableSync      bool   `json:"enable_sync"`
	SyncIntervalMs  uint32 `json:"sync_interval_ms"`
	MiningThreads   uint32 `json:"mining_threads"`
	MaxConnections  uint32 `json:"max_connections"`
	RPCPort         uint16 `json:"rpc_port"`
	P2PPort         uint16 `json:"p2p_port"`
	EnableRPC       bool   `json:"enable_rpc"`
	EnableP2P       bool   `json:"enable_p2p"`
	LogLevel        string `json:"log_level"`
	LogFile         string `json:"log_file"`
}

// DefaultConfig returns the default configuration
func DefaultConfig() *Config {
	return &Config{
		DataDir:         "./data",
		Network:         "mainnet",
		EnableMining:    false,
		EnableSync:      true,
		SyncIntervalMs:  1000,
		MiningThreads:   1,
		MaxConnections:  10,
		RPCPort:         7777,
		P2PPort:         67777,
		EnableRPC:       true,
		EnableP2P:       true,
		LogLevel:        "info",
		LogFile:         "./logs/satox-sdk.log",
	}
}

// NewSatoxSDK creates a new SDK instance
func NewSatoxSDK() *SatoxSDK {
	ctx, cancel := context.WithCancel(context.Background())
	return &SatoxSDK{
		config: DefaultConfig(),
		ctx:    ctx,
		cancel: cancel,
	}
}

// Initialize initializes the SDK with the given configuration
func (sdk *SatoxSDK) Initialize(config *Config) error {
	sdk.mutex.Lock()
	defer sdk.mutex.Unlock()

	if sdk.isRunning {
		return fmt.Errorf("SDK is already running")
	}

	if config != nil {
		sdk.config = config
	}

	// Initialize core manager
	sdk.coreManager = NewCoreManager()
	if err := sdk.coreManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize core manager: %w", err)
	}

	// Initialize wallet manager
	sdk.walletManager = NewWalletManager()
	if err := sdk.walletManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize wallet manager: %w", err)
	}

	// Initialize security manager
	sdk.securityManager = NewSecurityManager()
	if err := sdk.securityManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize security manager: %w", err)
	}

	// Initialize asset manager
	sdk.assetManager = NewAssetManager()
	if err := sdk.assetManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize asset manager: %w", err)
	}

	// Initialize NFT manager
	sdk.nftManager = NewNFTManager()
	if err := sdk.nftManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize NFT manager: %w", err)
	}

	// Initialize blockchain manager
	sdk.blockchainManager = NewBlockchainManager()
	if err := sdk.blockchainManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize blockchain manager: %w", err)
	}

	// Initialize IPFS manager
	sdk.ipfsManager = NewIPFSManager()
	if err := sdk.ipfsManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize IPFS manager: %w", err)
	}

	// Initialize network manager
	sdk.networkManager = NewNetworkManager()
	if err := sdk.networkManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize network manager: %w", err)
	}

	// Initialize database manager
	sdk.databaseManager = NewDatabaseManager()
	if err := sdk.databaseManager.Initialize(sdk.config); err != nil {
		return fmt.Errorf("failed to initialize database manager: %w", err)
	}

	sdk.isRunning = true
	return nil
}

// Start starts the SDK services
func (sdk *SatoxSDK) Start() error {
	sdk.mutex.Lock()
	defer sdk.mutex.Unlock()

	if !sdk.isRunning {
		return fmt.Errorf("SDK is not initialized")
	}

	// Start core manager
	if err := sdk.coreManager.Start(); err != nil {
		return fmt.Errorf("failed to start core manager: %w", err)
	}

	// Start other managers
	if err := sdk.walletManager.Start(); err != nil {
		return fmt.Errorf("failed to start wallet manager: %w", err)
	}

	if err := sdk.securityManager.Start(); err != nil {
		return fmt.Errorf("failed to start security manager: %w", err)
	}

	if err := sdk.assetManager.Start(); err != nil {
		return fmt.Errorf("failed to start asset manager: %w", err)
	}

	if err := sdk.nftManager.Start(); err != nil {
		return fmt.Errorf("failed to start NFT manager: %w", err)
	}

	if err := sdk.blockchainManager.Start(); err != nil {
		return fmt.Errorf("failed to start blockchain manager: %w", err)
	}

	if err := sdk.ipfsManager.Start(); err != nil {
		return fmt.Errorf("failed to start IPFS manager: %w", err)
	}

	if err := sdk.networkManager.Start(); err != nil {
		return fmt.Errorf("failed to start network manager: %w", err)
	}

	if err := sdk.databaseManager.Start(); err != nil {
		return fmt.Errorf("failed to start database manager: %w", err)
	}

	return nil
}

// Stop stops the SDK services
func (sdk *SatoxSDK) Stop() error {
	sdk.mutex.Lock()
	defer sdk.mutex.Unlock()

	if !sdk.isRunning {
		return nil
	}

	// Stop all managers
	sdk.coreManager.Stop()
	sdk.walletManager.Stop()
	sdk.securityManager.Stop()
	sdk.assetManager.Stop()
	sdk.nftManager.Stop()
	sdk.blockchainManager.Stop()
	sdk.ipfsManager.Stop()
	sdk.networkManager.Stop()
	sdk.databaseManager.Stop()

	sdk.isRunning = false
	return nil
}

// Shutdown gracefully shuts down the SDK
func (sdk *SatoxSDK) Shutdown() error {
	sdk.cancel()
	return sdk.Stop()
}

// IsRunning returns whether the SDK is running
func (sdk *SatoxSDK) IsRunning() bool {
	sdk.mutex.RLock()
	defer sdk.mutex.RUnlock()
	return sdk.isRunning
}

// GetCoreManager returns the core manager
func (sdk *SatoxSDK) GetCoreManager() *CoreManager {
	return sdk.coreManager
}

// GetWalletManager returns the wallet manager
func (sdk *SatoxSDK) GetWalletManager() *WalletManager {
	return sdk.walletManager
}

// GetSecurityManager returns the security manager
func (sdk *SatoxSDK) GetSecurityManager() *SecurityManager {
	return sdk.securityManager
}

// GetAssetManager returns the asset manager
func (sdk *SatoxSDK) GetAssetManager() *AssetManager {
	return sdk.assetManager
}

// GetNFTManager returns the NFT manager
func (sdk *SatoxSDK) GetNFTManager() *NFTManager {
	return sdk.nftManager
}

// GetBlockchainManager returns the blockchain manager
func (sdk *SatoxSDK) GetBlockchainManager() *BlockchainManager {
	return sdk.blockchainManager
}

// GetIPFSManager returns the IPFS manager
func (sdk *SatoxSDK) GetIPFSManager() *IPFSManager {
	return sdk.ipfsManager
}

// GetNetworkManager returns the network manager
func (sdk *SatoxSDK) GetNetworkManager() *NetworkManager {
	return sdk.networkManager
}

// GetDatabaseManager returns the database manager
func (sdk *SatoxSDK) GetDatabaseManager() *DatabaseManager {
	return sdk.databaseManager
}

// GetConfig returns the current configuration
func (sdk *SatoxSDK) GetConfig() *Config {
	return sdk.config
}

// Version returns the SDK version
func (sdk *SatoxSDK) Version() string {
	return "1.0.0"
}

// BuildDate returns the SDK build date
func (sdk *SatoxSDK) BuildDate() string {
	return "2025-06-30"
}

// Connect establishes a connection to the blockchain network
func (sdk *SatoxSDK) Connect() error {
	if !sdk.isRunning {
		return fmt.Errorf("SDK is not running")
	}
	return sdk.networkManager.Connect()
}

// Disconnect disconnects from the blockchain network
func (sdk *SatoxSDK) Disconnect() error {
	if !sdk.isRunning {
		return nil
	}
	return sdk.networkManager.Disconnect()
}

// WaitForShutdown waits for the SDK to be shut down
func (sdk *SatoxSDK) WaitForShutdown() {
	<-sdk.ctx.Done()
}

// WaitForShutdownWithTimeout waits for the SDK to be shut down with a timeout
func (sdk *SatoxSDK) WaitForShutdownWithTimeout(timeout time.Duration) error {
	select {
	case <-sdk.ctx.Done():
		return nil
	case <-time.After(timeout):
		return fmt.Errorf("shutdown timeout exceeded")
	}
} 